astropy/table/table.py

```diff
@@ -3963,7 +3963,7 @@ def _convert_col_for_table(self, col):
             # Quantity subclasses identified in the unit (such as u.mag()).
             q_cls = Masked(Quantity) if isinstance(col, MaskedColumn) else Quantity
             try:
-                qcol = q_cls(col.data, col.unit, copy=False, subok=True)
+                qcol = q_cls(col.data, col.unit, copy=False, subok=True, dtype=col.dtype)
```

Comment:
Looks fine but this should have an explicit test (even if it is implicitly tested in the mixin handling stuff).

Reply:
Sure; was just waiting whether to test `QTable(tab)` or only `QTable(tab, dtype=tab.dtype)`.

Reply:
Added under `TestSetTableColumn`